File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/screen/main/GistsListViewHolder.java
package ru.avystavkin.githubgists.screen.main;

import android.view.View;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import butterknife.BindView;
import ru.avystavkin.githubgists.AppDelegate;
import ru.avystavkin.githubgists.R;
import ru.avystavkin.githubgists.models.local.Gist;
import ru.avystavkin.githubgists.screen.base.holders.BaseViewHolder;
import ru.avystavkin.githubgists.screen.interfaces.OnGistClickListner;
import ru.avystavkin.githubgists.screen.interfaces.OnItemClickListener;
import ru.avystavkin.githubgists.screen.main.gist.GistsAdapter;
import ru.avystavkin.githubgists.widget.DividerItemDecoration;
import ru.avystavkin.githubgists.widget.EmptyRecyclerView;

public class GistsListViewHolder extends BaseViewHolder implements OnItemClickListener<Gist> {

    @BindView(R.id.recyclerViewVertical)
    EmptyRecyclerView mRecyclerView;

    @BindView(R.id.empty)
    View mEmptyView;

    @Nullable
    private OnGistClickListner mOnGistClickListener;

    private GistsAdapter mAdapter;

    public GistsListViewHolder(@Nullable OnGistClickListner onGistClickListener, View itemView) {
        super(itemView);

        mRecyclerView.setLayoutManager(new LinearLayoutManager(AppDelegate.getContext()));
        mRecyclerView.addItemDecoration(new DividerItemDecoration(AppDelegate.getContext()));
        mRecyclerView.setEmptyView(mEmptyView);

        mAdapter = new GistsAdapter(new ArrayList<>());
        mAdapter.attachToRecyclerView(mRecyclerView);
        mAdapter.setOnItemClickListener(this);

        mOnGistClickListener = onGistClickListener;
    }

    public void bind(@NonNull List<Gist> gists) {
        mAdapter.changeDataSet(gists);
    }

    @Override
    public void onItemClick(@NonNull Gist gist) {
        if (mOnGistClickListener != null)
            mOnGistClickListener.onGistClick(gist);
    }
}

File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/screen/main/GistsListViewHolder.kt
package ru.avystavkin.githubgists.screen.main

import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import butterknife.BindView
import ru.avystavkin.githubgists.AppDelegate
import ru.avystavkin.githubgists.R
import ru.avystavkin.githubgists.models.local.Gist
import ru.avystavkin.githubgists.screen.base.holders.BaseViewHolder
import ru.avystavkin.githubgists.screen.interfaces.OnGistClickListner
import ru.avystavkin.githubgists.screen.interfaces.OnItemClickListener
import ru.avystavkin.githubgists.screen.main.gist.GistsAdapter
import ru.avystavkin.githubgists.widget.DividerItemDecoration
import ru.avystavkin.githubgists.widget.EmptyRecyclerView
import java.util.*

class GistsListViewHolder(private val mOnGistClickListener: OnGistClickListner?, itemView: View) : BaseViewHolder(itemView), OnItemClickListener<Gist> {

    @BindView(R.id.recyclerViewVertical)
    internal lateinit var mRecyclerView: EmptyRecyclerView

    @BindView(R.id.empty)
    internal lateinit var mEmptyView: View

    private val mAdapter: GistsAdapter

    init {

        mRecyclerView.layoutManager = LinearLayoutManager(AppDelegate.context)
        mRecyclerView.addItemDecoration(DividerItemDecoration(AppDelegate.context))
        mRecyclerView.emptyView = mEmptyView

        mAdapter = GistsAdapter(ArrayList())
        mAdapter.attachToRecyclerView(mRecyclerView)
        mAdapter.onItemClickListener = this
    }

    fun bind(gists: List<Gist>) {
        mAdapter.changeDataSet(gists)
    }

    override fun onItemClick(gist: Gist) {
        mOnGistClickListener?.onGistClick(gist)
    }
}
